lib/matplotlib/dates.py

```diff
@@ -430,6 +430,11 @@ def date2num(d):
     For details see the module docstring.
     """
 
+    if hasattr(d, "values"):
+        d = d.values
+    if hasattr(d, "to_pydatetime"):
+        d = d.to_pydatetime()
```

Comment:
I am not sure this one is needed, as `date2num` already handles Timestamp objects (them being datetime.datetime subclasses):

```
In [82]: date2num(pd.Timestamp('2012-01-01'))
Out[82]: 734503.0

In [85]: date2num(datetime.datetime(2012, 1, 1))
Out[85]: 734503.0
```